projects/plugins/jetpack/modules/google-fonts/class-jetpack-google-fonts.php

```diff
@@ -0,0 +1,193 @@
+
+		/**
+		 * We are already enqueueing all registered fonts by default when loading the block editor,
+		 * so we only need to scan for webfonts when browsing as a guest.
+		 */
+		if ( ! is_admin() ) {
+			add_action( 'wp_loaded', array( $this, 'scan_global_styles_for_google_fonts' ) );
+			add_filter( 'pre_render_block', array( $this, 'scan_block_for_google_fonts' ), 10, 2 );
+		}
+	}
+
+	/**
+	 * Register the provider for Google fonts registration.
+	 */
+	private function register_google_fonts_provider() {
+		if ( ! function_exists( 'wp_register_webfont_provider' ) ) {
+			return;
+		}
+
+		wp_register_webfont_provider( 'jetpack-google-fonts', '\Automattic\Jetpack\Fonts\Google_Fonts_Provider' );
+	}
+
+	/**
+	 * Register Google fonts.
+	 */
+	public function register_google_fonts() {
+		if ( ! function_exists( 'wp_register_webfonts' ) ) {
+			return;
+		}
+
+		/**
+		 * Curated list of Google Fonts.
+		 *
+		 * @module google-fonts
+		 *
+		 * @since 10.8
+		 *
+		 * @param array $fonts_to_register Array of Google Font names to register.
+		 */
+		$fonts_to_register = apply_filters( 'jetpack_google_fonts_list', JETPACK_GOOGLE_FONTS_LIST );
+
+		foreach ( $fonts_to_register as $font_family ) {
+			wp_register_webfonts(
+				array(
+					array(
+						'font-family'  => $font_family,
+						'font-weight'  => '100 900',
+						'font-style'   => 'normal',
+						'font-display' => 'fallback',
+						'provider'     => 'jetpack-google-fonts',
+					),
+					array(
+						'font-family'  => $font_family,
+						'font-weight'  => '100 900',
+						'font-style'   => 'italic',
+						'font-display' => 'fallback',
+						'provider'     => 'jetpack-google-fonts',
+					),
+				)
+			);
+
+			$slug = wp_webfonts()->get_font_slug( $font_family );
+
+			/**
+			 * When introspecting, all we have is the slug,
+			 * so we need to keep track of the registered families
+			 * to check whether the font was filtered after
+			 * running the `jetpack_google_fonts_list` hook.
+			 */
+			$this->registered_google_fonts[ $slug ] = $font_family;
+		}
+	}
+
+	/**
+	 * Scan block for Google fonts.
+	 *
+	 * @param string $content The block content.
+	 * @param array  $parsed_block The parsed block attributes.
+	 *
+	 * @return array
+	 */
+	public function scan_block_for_google_fonts( $content, $parsed_block ) {
+		if ( isset( $parsed_block['attrs']['fontFamily'] ) ) {
+			$this->maybe_enqueue_font_family( $parsed_block['attrs']['fontFamily'] );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Enqueue a font family if it was not removed from the curated list.
+	 *
+	 * @param string $font_family_slug The block content.
+	 */
+	private function maybe_enqueue_font_family( $font_family_slug ) {
+		if ( ! isset( $this->registered_google_fonts[ $font_family_slug ] ) ) {
+			// Font not allow-listed.
+			return;
+		}
+
+		wp_enqueue_webfont( $font_family_slug );
+	}
+
+	/**
+	 * Scan global styles for Google fonts.
+	 */
+	public function scan_global_styles_for_google_fonts() {
```

Comment:
This function probably does not belong here. Right now, however, I'm struggling to find a better place to put it...

When it's time, we can probably move this to Gutenberg.

Reply:
I wonder if this could be in the package for now, maybe in the Utils class?
https://github.com/Automattic/jetpack/tree/212b4d7bc0ab2be67b0ed7347f6496454e390052/projects/packages/google-fonts-provider/src

Reply:
I actually created a brand new package. It didn't feel connected to the provider at all, and it could be used by other modules so hurray for reusability!

Reply:
I wonder if this may be overkill. When used by third-parties, wouldn't that code pretty much always be used alongside the existing `automattic/jetpack-google-fonts-provider` package? Even if not fully related, maybe it could be a new class inside that package? We could always extract it into a new package if we start needing it elsewhere in the future, somewhere where the jetpack-google-fonts-provider package is not already required.

Reply:
> I wonder if this may be overkill.

I would agree; I think separating this out into its own package doesn't really provide much benefit. Realistically, you should always be introspecting fonts when using the provider with more than a few font faces (especially with how Google creates so many font face declarations for different character ranges). I'd hope that packaging the provider and introspection utilities together would encourage that.

If `automattic/jetpack-google-fonts-provider ` as a package is too specific, I'd rather deprecate it and create a new package to more broadly encompass webfont functionality.

Additionally, having Global styles introspection and block introspection in separate places seems odd; I think these should be together.

Reply:
Yep. Any reason why we created this `jetpack-google-fonts-provider` package in the first place? It looks weird to have this as separate parts.

And I do agree -- I think it's better if we collocated the provider and the introspector together within the Jetpack Google Fonts module. Will do that.

Reply:
Sorry if I wasn't clear, I don't mean stop using packages all together. My understanding is that Jetpack (the monorepo) is built around packages for functionality that's reusable across plugins. That way if Woo, for example, wants to use something that's in Jetpack, it doesn't have to copy the code into itself, or require install the entire Jetpack plugin be installed, it can just require the specific package it needs.

So for our use case, I think this separation makes sense:

- Package: font provider, and any functionality needed for loading fonts from the provider. No hooks should be added by default, so it's up to package consumers to "turn on" any functionality from the package they want to use.
- Module: Our WP.com specific list of curated Google Fonts, and calling the necessary functions from the package to load those fonts.

The main goal of the separation here being any other plugin could use the fonts package to load it's own set of Google Fonts, independent from the fonts module in the Jetpack plugin.

Reply:
So, a few things:

- The global styles introspection utils package is not adding any hook -- it's simply getting the global styles and looking for font families. However, that might be incomplete because only global styles introspection is not enough even for this use case as block settings are ignored. One thing I would like to have is an action called `jetpack_content_introspection_webfont_found` or something like that would get fired whenever a font family was found in the content;
- The introspection package is not really related to the provider. Adding the provider does not add introspection, and vice-versa. One could use introspection without necessarily working with Google provider. The provider's job is to handle and emit the enqueued fonts, and that's it. It doesn't care about introspection.
- The module could (and will) use both packages. Other modules might not, and only use the provider, or only use the introspection... I do think they should be decoupled as those are very useful things to expose.

Reply:
> One could use introspection without necessarily working with Google provider. The provider's job is to handle and emit the enqueued fonts, and that's it. It doesn't care about introspection.

Imagining myself as a developer with 0 knowledge of the fonts work we've done, I'd find it unintuitive to need to import two separate packages to enqueue webfonts in a performant way. That is to say, I wouldn't think to look for an introspector while setting up fonts support. 

I agree with Jeremy + Grant -- I suspect there'd be more clarity in colocating both introspection and provider logic in the same package for now.